deployment/data-streams/utils/mcmsutil/mcmsutil.go

```diff
@@ -90,3 +92,76 @@ func ExecuteOrPropose(
 	_, err := txutil.SignAndExecute(e, txs)
 	return deployment.ChangesetOutput{}, err
 }
+
+// TransferToMCMSWithTimelockForTypeAndVersion transfers ownership of the contracts of a specific type and version to the
+// MCMS timelock on that chain. The output will contain an MCMS timelock proposal for "AcceptOwnership" of those contracts
+// The address book should be recently deployed addresses that are being transferred to MCMS and should not be in e.ExistingAddresses
+func TransferToMCMSWithTimelockForTypeAndVersion(e deployment.Environment, ab deployment.AddressBook,
+	filter deployment.TypeAndVersion, mcmsConfig proposalutils.TimelockConfig) (deployment.ChangesetOutput, error) {
+	contractAddresses := make(map[uint64][]common.Address)
+	for _, chain := range e.Chains {
+		chainAddresses, err := ab.AddressesForChain(chain.Selector)
+		if err != nil {
+			return deployment.ChangesetOutput{}, fmt.Errorf("failed to get addresses from address book: %w", err)
+		}
+		for address, typeAndVersion := range chainAddresses {
+			if typeAndVersion.Type == filter.Type && typeAndVersion.Version == filter.Version {
+				contractAddresses[chain.Selector] = append(contractAddresses[chain.Selector], common.HexToAddress(address))
+			}
+		}
+	}
+
+	// create a merged addressbook with the existing + new addresses. Sub-changesets will need all addresses
+	// This is required when chaining together changesets
+	existingAddresses := e.ExistingAddresses
+	abTemp := deployment.NewMemoryAddressBook()
+	if err := abTemp.Merge(ab); err != nil {
+		return deployment.ChangesetOutput{}, fmt.Errorf("failed merging new addresses into temp addresses: %w", err)
+	}
+	if err := abTemp.Merge(e.ExistingAddresses); err != nil {
+		return deployment.ChangesetOutput{}, fmt.Errorf("failed merging existing addresses into temp addresses: %w", err)
+	}
+
+	e.ExistingAddresses = abTemp
+
+	transferCs := deployment.CreateLegacyChangeSet(commonchangeset.TransferToMCMSWithTimelockV2)
+	transferCfg := commonchangeset.TransferToMCMSWithTimelockConfig{
+		ContractsByChain: contractAddresses,
+		MCMSConfig:       mcmsConfig,
+	}
+
+	transferOut, err := transferCs.Apply(e, transferCfg)
+	e.ExistingAddresses = existingAddresses // reset the address book to the original state regardless of errors
```

Comment:
Purely out of curiosity, why do we need to reset the env's AB? We are not using the env for anything else here.

Reply:
Can be removed since I've introduced the env cloning